system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1276,416 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+          gboolean dir_in_use = FALSE;
+
+          /* If this cache directory is used by any other ongoing pull, we
+           * should not be re(using) it. */
+          G_LOCK (ongoing_pulls);
+          for (guint i = 0; i < ongoing_pulls->len; i++)
+            {
+              OngoingPull *pull = g_ptr_array_index (ongoing_pulls, i);
+              if (g_strcmp0 (pull->src_dir, name) == 0)
+                {
+                  dir_in_use = TRUE;
+                  break;
+                }
+            }
+          G_UNLOCK (ongoing_pulls);
+
+          if (!dir_in_use && check_for_repo_containing_partial_pull (cache_dir, commit, out_location, passwd))
+            return TRUE;
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+handle_get_revokefs_fd (FlatpakSystemHelper   *object,
+                        GDBusMethodInvocation *invocation,
+                        GUnixFDList           *arg_fdlist,
+                        guint                  arg_flags,
+                        const gchar           *arg_ref,
+                        const gchar           *arg_commit,
+                        const gchar           *arg_installation)
+{
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GUnixFDList) fd_list = NULL;
+  g_autoptr(GError) error = NULL;
+  g_auto(GStrv) parts = NULL;
+  g_autofree gchar *src = NULL;
+  g_autofree gchar *flatpak_dir = NULL;
+  g_autofree gchar *repo_tmp = NULL;
+  g_autofree gchar *cache_dir_basename = NULL;
+  g_autofree gchar *partial_pull_repo = NULL;
+  g_autofree gchar *passwd_buf = NULL;
+  struct passwd passwd;
+  OngoingPull *new_pull;
+  uid_t uid;
+  int fd_index = -1;
+
+  g_debug ("GetRevokefsFd %u %s %s %s", arg_flags, arg_ref, arg_commit, arg_installation);
+
+  system = dir_get_system (arg_installation, get_sender_pid (invocation), &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if ((arg_flags & ~FLATPAK_HELPER_GET_REVOKEFS_FD_FLAGS_ALL) != 0)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_INVALID_ARGS,
+                                             "Unsupported flags enabled: 0x%x", (arg_flags & ~FLATPAK_HELPER_GET_REVOKEFS_FD_FLAGS_ALL));
+      return TRUE;
+    }
+
+  if (!check_for_system_helper_user (&passwd, &passwd_buf, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if (!get_connection_uid (invocation, &uid, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+  repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+  parts = flatpak_decompose_ref (arg_ref, NULL);
+
+   if (check_partial_pull_availability (repo_tmp, parts[1], arg_commit, &partial_pull_repo, &passwd))
```

Comment:
There seems to be a race here. Nothing prohibits two callers to chose the same existing directory if they both run this line before we add it to the ongoing_pulls array.

Reply:
In addition, it seems to me that it can race against handle_deploy(). I.e. that takes the directory out of the array, but what prohibits another thread to then try to re-use that for a new pull?

Reply:
Yes. There is a potential race.

To address this, I tweak-ed the implementation to use a GHashTable instead of GPtrArray.
The keys in the hash-table are cache-directories source paths. The value can be NULL or (OngoingPull *).

If the source-path(as key) is present in the hash-table, it means the directory is in use. 
The value point to a `OngoingPull *` means the revokefs-fuse machinery is setup and running.
The value denoting `NULL` means the revokefs-machinery is **yet** to setup or directory is being deployed. Does this approach address the issue above? 

071a00f57a9ac is the working implementation of the above approach (the needs to be reviewed). 
(Thanks @wjt for the suggestion of the approach).